README
                      ---------------------------------
                      SMALL-FOOTPRINT CIM BROKER (sfcb)
                      ---------------------------------

What is sfcb
==============================
sfcb is a lightweight CIM daemon (aka CIMOM) that responds to CIM client
requests for system management data and/or performs system management 
tasks. sfcb supports most of the standard CIM XML over http/https protocol.
It is highly modular, allowing functionality to be easily added, removed
or customized for different management applications. sfcb is specifically
targeted for small embedded system that do not have the available CPU,
memory or disk resources to support a full-blown enterprise-level CIMOM.
That said, sfcb runs very well on a regular Linux/Unix system and supports
most of the functions required by CIM clients to manage such the system. 

Documentation
==============================

The sfcb documentation is maintained in the sfcb wiki on Sourceforge.

sfcb wiki: http://sourceforge.net/apps/mediawiki/sblim/index.php?title=Sfcb
sfcb documentation: https://sourceforge.net/apps/mediawiki/sblim/index.php?title=SfcbTheBook

Quick Install
==============================

The following procedure will get SFCB installed with default options. If you 
need to customize the installation at all, please see the documentation
linked to above.

    The following prerequisites must be installed:

        sfcCommon
        cmpi-devel >= 2.0.1
        automake >= 1.8
        autoconf >= 2.5.8
        libtool >= 1.5.8
        bison >= 1.85
        flex
        curl >= 7.11.1-1
        curl-devel >= 7.11.1-1
    

    3. Build sfcb using the following commands from the sfcb directory:

        ./autoconfiscate.sh (if you obtained the source from CVS/GIT)
        ./configure
        make
        sudo make install
        sudo make postinstall

    4. Start sfcbd:

        sudo /usr/local/sbin/sfcbd
    

Support and Reporting Bugs
==============================
The SBLIM project maintains a mailing list to which you should direct all
questions regarding sfcb, sfcc, wbemcli, or any of the SBLIM providers.
To subscribe to this mailing list visit

        http://lists.sourceforge.net/lists/listinfo/sblim-devel

The SBLIM Project Bug tracking page can be found at:

       http://sourceforge.net/tracker/?group_id=128809&atid=712784

Here you will find a list of all known bugs for SFCB (set the Category
field to "sfcb").  You can also report new bugs and request features.
When reporting a bug, please provide a good description and attach any
useful files, such as patches or the XML used to generate the error




NEWS
Changes in 1.4.10
=================

New features:

Bugs fixed:


Changes in 1.4.9
================

New features:
- [sfcb-tix:#97] Add a select timeout for req handlers

Bugs fixed:
- [sfcb-tix:#106] CMPIRole not passed to provider context
- [sfcb-tix:#107] Quietly unpack schema at make postinstall
- [sfcb-tix:#108] Some associatorname CIM operations hang

Changes in 1.4.8
================

New features:
- [sfcb-tix:#99] Add config property enableSslCipherServerPref
- [sfcb-tix:#101] Add config property indicationCurlUseExpect100
- [sfcb-tix:#102] Add config property indicationCurlHonorRedirect
- [sfcb-tix:#105] Add config property embeddedObjectEncoding

Bugs fixed:
- [sfcb-tix:#94] Increase sigChldWaitTime
- [sfcb-tix:#95] Do not release pthread lock prior to stopBroker() exit
- [sfcb-tix:#89] Gracefully handle providers doing abort()
- [sfcb-tix:#96] Provider should close req handler on shutdown
- [sfcb-tix:#98] sfcbmofpp segfaults if mof file ends with block comment without newline
- [sfcb-tix:#100] Fix some compiler warnings & errors on newer distros
- [sfcb-tix:#103] Unsafe usage of strncpy in providerRegister.c

Changes in 1.4.7
================

New features:
- [sfcb-tix:#86] Support configurable SSL Diffie Hellman parameters file
- [sfcb-tix:#87] Support configurable SSL ECDH elliptic curve name
- [sfcb-tix:#76] Identify running SFCB processes in ps output

Bugs fixed:
- [sfcb-tix:#81] Bad CMPIType assignment in cimXmlOps
- [sfcb-tix:#82] ClientEnvFT for SfcbLocal doesn't match SFCC
- [sfcb-tix:#83] Add Alternate getObjectPath() for SfcbLocal Connections
- [sfcb-tix:#84] Upcall mutex not freed on shutdown
- [sfcb-tix:#85] SFCB: SfcbLocal interface should expose markHeap() and releaseHeap()
- [sfcb-tix:#88] Ensure unique semaphore keys across multiple http adapters
- [sfcb-tix:#92] indicationDeliveryThreadLimit not properly enforced
- [sfcb-tix:#93] ProfileProvider segfault while stopping slpUpdate thread
- [sfcb-tix:#90] Problems with enableInteroOp=false

Changes in 1.4.6
================

New features:
- [sfcb-tix:#64] Support maxObjCount for HTTP chunking
- [sfcb-tix:#70] Reload SSL context-without restarting httpd

Bugs fixed:
- [sfcb-tix:#61] classProviderSf.c SIGSEGV if repo is not initialized
- [sfcb-tix:#63] Improve NamespaceProvider and ClassProvider Consistency
- [sfcb-tix:#66] SFCB_TRACE_VAR should be more restrictive
- [sfcb-tix:#60] fix providerMgr trace for localconnect
- [sfcb-tix:#68] reinitialize select timeout in SSL handshake
- [sfcb-tix:#79] sfcc cannot reconnect with sfcb if sfcb restarts [bugs:#2663]
- [sfcb-tix:#69] Fix logger for long-lived clients, was: "#2347 Don't spawn logger for clients"
- [sfcb-tix:#72] Make systemddir configurable
- [sfcb-tix:#74] providerRegister: check for and prevent buffer overflows
- [sfcb-tix:#75] fileRepository: show which directory could not be opened
- [sfcb-tix:#78] Missing null check in ift_getObjectPath, transposed debug stmts in memAlloc
- [sfcb-tix:#78] Simplify mutex handling in instance.c
- [sfcb-tix:#77] Fix prototype for stopBroker
- [sfcb-tix:#80] sfcc cannot reconnect with sfcb

Changes in 1.4.5
================

New features:
- [sfcb-tix:#44] Escape XML for Embedded Instances Instead of Using CDATA
- [sfcb-tix:#48] mofc: Allow absoltute paths in #pragma include
- [sfcb-tix:#52] ABT core check
- [sfcb-tix:#55] make install should not overwrite sfcb.cfg
- [sfcb-tix:#59] Improve httpAdapter error handling

Bugs fixed:
- [sfcb-tix:#31] TestAssociationProvider can cause some queries to fail
- [sfcb-tix:#32] Provider Mgr trace should id the provider being called
- [sfcb-tix:#36] Add Pulled Enums test provider
- [sfcb-tix:#38] providerRegister:locateProvider() is unused
- [sfcb-tix:#40] Add valgrind test (implemented only for the "none" init script at this time)
- [sfcb-tix:#41] Global IndicationService
- [sfcb-tix:#46] Error parsing CDATA with no whitespace
- [sfcb-tix:#47] Pulled Enumeration support - CIMXML parser
- [sfcb-tix:#37] indCIMXmlHandler coredumps in IndCIMXMLHandlerInvokeMethod
- [sfbb-tix:#51] Malformed OP causes segfault
- [sfcb-tix:#49] Cannot return EmbeddedInstance property from provider
- [sfcb-tix:#53] SFCB may not start if runtime does not support IPv6
- [sfcb-tix:#54] Fix for sfcbproc.c
- [sfcb-tix:#18] disable default properties breaks test. Modified failing tests, also made test case insensitive and fixed result file naming.
- [sfcb-tix:#56] Intermittent segfault at shutdown Relocate trace_stop due to timing issue on shmdt() call
- [sfcb-tix:#12] Support HTTP Expect: 100-continue header
- [sfcb-tix:#11] Remove unnecessary casts for allocs
- [sfcb-tix:#58] sfcbdump test fails. Remove local test classSchemas file and use installed one instead.

Changes in 1.4.4
================

New features:
- 3597805 Restart HTTP daemon without restarting SFCB
- 3597806 Bind to arbitrary list of IP addresses in SFCB
- 3601386 Add test for EmbeddedObject output
- 3599160 Add sfcbproc utility
- 3601943 ObjectImpl unit test
- 2766931 Property filters for interop and interopserver providers
- [sfcb-tix:#16] getClass does not expose method information for localconnect

Bugs fixed:
- 3599526 segfault during ecn with classProviderSf
- 3599329 dlopen failed for sfcbCustomLib
- 3600125 Share client tracing shm segment
- 3600283 Fixup compiler warnings from gcc 4.7
- 3600159 TestProviders should have their own Makefile
- 3601026 Fixup compiler warnings from -Wextra
- 3599437 missing include of <sys/resource.h>
- 3601301 Add stubs for empty CMPI functions for internal providers
- 2878328 sfcb and mofc won't compile with --disable-qualifierrep
- 3601661 Reenable no-fork http adapter mode
- 2990304 Enhancements for sfcb
- 2778345 There is no response at GetClass after CreateClass
- 3603454 control should do type conversions on startup
- [Bugs:#2612] Don't persist indications
- [Bugs:#2613] Check sfcCommon version
- [sfcb-tix:#21] datetime conversions do not handle timezone

trace.c

/*
 * trace.c
 *
 * (C) Copyright IBM Corp. 2005
 *
 * THIS FILE IS PROVIDED UNDER THE TERMS OF THE ECLIPSE PUBLIC LICENSE
 * ("AGREEMENT"). ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS FILE
 * CONSTITUTES RECIPIENTS ACCEPTANCE OF THE AGREEMENT.
 *
 * You can obtain a current copy of the Eclipse Public License from
 * http://www.opensource.org/licenses/eclipse-1.0.php
 *
 * Author:        Adrian Schuur <schuur@de.ibm.com>
 * Based on concepts developed by Heidi Neumann <heidineu@de.ibm.com>
 *
 * Description:
 *
 * Trace support for sfcb.
 *
 */

#include "trace.h"
#include <errno.h>
#include "native.h"
#include <string.h>
#include <time.h>

#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/types.h>
#include <unistd.h>
#include <pthread.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <stdlib.h>
#include "config.h"

/*
 * ---------------------------------------------------------------------------
 */
/*
 * trace facility 
 */
/*
 * ---------------------------------------------------------------------------
 */

char           *processName = NULL;
int             providerProcess = 0;
int             idleThreadId = 0;
int             terminating = 0;

int             _sfcb_debug = 0;
unsigned long   _sfcb_trace_mask = 0;
/* use pointer indirect _sfcb_trace_mask to allow shared memory flag */
unsigned long *_ptr_sfcb_trace_mask = &_sfcb_trace_mask;
void *vpDP = NULL;
int shmid=0;
char           *_SFCB_TRACE_FILE = NULL;
int _SFCB_TRACE_TO_SYSLOG = 0;

TraceId         traceIds[] = {
  {"providerMgr", TRACE_PROVIDERMGR},
  {"providerDrv", TRACE_PROVIDERDRV},
  {"cimxmlProc", TRACE_CIMXMLPROC},
  {"httpDaemon", TRACE_HTTPDAEMON},
  {"upCalls", TRACE_UPCALLS},
  {"encCalls", TRACE_ENCCALLS},
  {"ProviderInstMgr", TRACE_PROVIDERINSTMGR},
  {"providerAssocMgr", TRACE_PROVIDERASSOCMGR},
  {"providers", TRACE_PROVIDERS},
  {"indProvider", TRACE_INDPROVIDER},
  {"internalProvider", TRACE_INTERNALPROVIDER},
  {"objectImpl", TRACE_OBJECTIMPL},
  {"xmlIn", TRACE_XMLIN},
  {"xmlOut", TRACE_XMLOUT},
  {"sockets", TRACE_SOCKETS},
  {"memoryMgr", TRACE_MEMORYMGR},
  {"msgQueue", TRACE_MSGQUEUE},
  {"xmlParsing", TRACE_XMLPARSING},
  {"responseTiming", TRACE_RESPONSETIMING},
  {"dbpdaemon", TRACE_DBPDAEMON},
  {"slp", TRACE_SLP},
  {NULL, 0}
};

/*
 * set the fg color on the console reset = 1 will restore previous color
 * set see trace.h for valid attr, fg, bg values
 * 
 * TODO: fix reset of fg color.  currently we assume it was white 
 */
void
changeTextColor(int reset)
{
  char            command[13];
  int             attr = (reset) ? RESET : currentProc % 2;
  int             fg = (reset) ? WHITE : currentProc % 7;
  int             bg = BLACK;
  if (fg == BLACK)
    fg = WHITE;
  sprintf(command, "%c[%d;%d;%dm", 0x1B, attr, fg + 30, bg + 40);
  fprintf(stderr, "%s", command);
}

void
_sfcb_trace_start(int n)
{
  if (_sfcb_debug < n) {
    _sfcb_debug = n;
  }
}

void
_sfcb_trace_stop()
{
  shmctl(shmid, IPC_RMID, 0);
  shmdt(vpDP);
  _sfcb_debug = 0;
}

void
_sfcb_trace_init()
{

  char           *var = NULL;
  char           *err = NULL;
  FILE           *ferr = NULL;
  int tryid = 0xDEB001;
  if (shmid == 0) {
    while ((shmid = shmget(tryid, sizeof(unsigned long), (IPC_EXCL | IPC_CREAT | 0660))) < 0 && (errno == EEXIST)) tryid++;
  }
  mlogf(M_INFO,M_SHOW,"--- Shared memory ID for tracing: %x\n", tryid);
  if (shmid < 0) {
    mlogf(M_ERROR,M_SHOW, "shmget(%x) failed in %s at line %d.\n", tryid, __FILE__, __LINE__ );
    abort();
  }
  else {
    vpDP = shmat( shmid, NULL, 0 );
	
    if (vpDP == (void*)-1) {// shmat returns an error
      mlogf(M_ERROR,M_SHOW, "shmat(%u,) failed with errno = %s(%u) in %s at line %d.\n", shmid, strerror(errno), errno, __FILE__, __LINE__ );
      abort();
    }
    else {
      _ptr_sfcb_trace_mask = (unsigned long *)vpDP;
    }
  }

  var = getenv("SFCB_TRACE");
  if (var != NULL) {
    _sfcb_debug = atoi(var);
  } else {
    _sfcb_debug = 0;
  }

  err = getenv("SFCB_TRACE_FILE");
  if (err != NULL) {
    if ((((ferr = fopen(err, "a")) == NULL) || fclose(ferr))) {
      mlogf(M_ERROR, M_SHOW, "--- Couldn't create trace file\n");
      return;
    }
    _SFCB_TRACE_FILE = strdup(err);
  } else {
    if (_SFCB_TRACE_FILE)
      free(_SFCB_TRACE_FILE);
    _SFCB_TRACE_FILE = NULL;
  }
}

char           *
_sfcb_format_trace(char *fmt, ...)
{
  va_list         ap;
  char           *msg = malloc(MAX_MSG_SIZE);
  va_start(ap, fmt);
  vsnprintf(msg, MAX_MSG_SIZE, fmt, ap);
  va_end(ap);
  return msg;
}

void
_sfcb_trace(int level, char *file, int line, char *msg)
{
  if (msg == NULL) return;

  struct tm       cttm;
  struct timeval  tv;
  struct timezone tz;
  long            sec = 0;
  char           *tm = NULL;
  FILE           *ferr = NULL;

  if ((_SFCB_TRACE_FILE != NULL)) {
    if ((ferr = fopen(_SFCB_TRACE_FILE, "a")) == NULL) {
      mlogf(M_ERROR, M_SHOW, "--- Couldn't open trace file");
      return;
    }
    colorTrace = 0;             /* prevents escape chars from getting into 
                                 * the file */
  } else {
    ferr = stderr;
  }

  if (gettimeofday(&tv, &tz) == 0) {
    sec = tv.tv_sec + (tz.tz_minuteswest * -1 * 60);
    tm = malloc(20 * sizeof(char));
    memset(tm, 0, 20 * sizeof(char));
    if (gmtime_r(&sec, &cttm) != NULL) {
      strftime(tm, 20, "%m/%d/%Y %H:%M:%S", &cttm);
    }

    if (*_ptr_sfcb_trace_mask) {
      if (_SFCB_TRACE_TO_SYSLOG) {
        /* ERROR is the default syslog level, if a user does not specify INFO or DEBUG. 
           ERROR guarantees output will end up in syslog */
        mlogf(M_ERROR,M_SHOW,"[%i] [%s] %d/%p --- %s(%i) : %s\n", level, tm, currentProc, (void *)pthread_self(), file,
              line, msg);
      }
      else if (colorTrace) {
        changeTextColor(0);
        fprintf(ferr, "[%i] [%s] %d/%p --- %s(%i) : %s\n", level, tm, currentProc, (void *)pthread_self(), file,
                line, msg);
        changeTextColor(1);
      }
      else {
        fprintf(ferr, "[%i] [%s] %d/%p --- %s(%i) : %s\n", level, tm, currentProc, (void *)pthread_self(), file,
                line, msg);
      }
    }

    free(tm);
    free(msg);

  }

  if ((_SFCB_TRACE_FILE != NULL)) {
    fclose(ferr);
  }
    
}

extern void _sfcb_set_trace_mask(unsigned long n)
{
  unsigned long *pulDP = (unsigned long*)vpDP;
  *pulDP = n;
}

extern void
_sfcb_set_trace_file(char *file)
{
  if (_SFCB_TRACE_FILE) {
    free(_SFCB_TRACE_FILE);
  }
  if (strcmp(file, "syslog") == 0) {
    _SFCB_TRACE_FILE = NULL;
    _SFCB_TRACE_TO_SYSLOG = 1;
  }
  else if (strcmp(file,"stderr") == 0) {
    _SFCB_TRACE_FILE = NULL;
  } else {
    _SFCB_TRACE_FILE = strdup(file);
  }
}

void
_sfcb_trap(int tn)
{
#ifdef SFCB_IX86
  char           *tp;
  int             t;
  if ((tp = getenv("SFCB_TRAP"))) {
    t = atoi(tp);
    if (tn == t)
      asm("int $3");
  }
#endif
}

sigHandler     *
setSignal(int sn, sigHandler * sh, int flags)
{
  struct sigaction newh,
                  oldh;
  newh.sa_handler = sh;
  sigemptyset(&newh.sa_mask);
  newh.sa_flags = flags;

  if (sn == SIGALRM)
    newh.sa_flags |= SA_INTERRUPT;
  else if (sn == SIGUSR2)
    newh.sa_flags |= SA_NODEFER;

  if (sigaction(sn, &newh, &oldh) < 0)
    return SIG_ERR;

  return oldh.sa_handler;
}

#ifdef UNITTEST
// Embedded unittest routine
int
trace_test()
{
  int             fail = 0;
  _sfcb_trace_start(1);
  if (_sfcb_debug < 1) {
    printf("  _sfcb_trace_start() test failed.\n");
    fail = 1;
  }
  _sfcb_trace_stop();
  if (_sfcb_debug != 0) {
    printf("  _sfcb_trace_stop() test failed.\n");
    fail = 1;
  }
  _sfcb_trace_init();
  if (_SFCB_TRACE_FILE == NULL) {
    printf("  _sfcb_trace_init() test failed.\n");
    fail = 1;
  }
  int             num = 99;
  char           *msg = _sfcb_format_trace("%s %d", "Test message", num);
  if (strcmp(msg, "Test message 99")) {
    printf("  _sfcb_format_trace() test failed, message=%s.\n", msg);
    fail = 1;
  }

  return fail;
}
#endif
/* MODELINES */
/* DO NOT EDIT BELOW THIS COMMENT */
/* Modelines are added by 'make pretty' */
/* -*- Mode: C; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
/* vi:set ts=2 sts=2 sw=2 expandtab: */
